tests/database/test_header_db.py

```diff
@@ -152,6 +207,111 @@ class StepAction(enum.Enum):
+            (StepAction.PERSIST_CHECKPOINT, 4),
+            (StepAction.PERSIST_HEADERS, ('a', lambda headers: headers[4:5])),
+            (StepAction.PERSIST_CHECKPOINT, 1),
+            (StepAction.VERIFY_PERSIST_RAISES, (
+                'b',
+                CheckpointsMustBeCanonical,
+                lambda headers: headers[4:6],
+            )),
+            (
+                StepAction.VERIFY_CANONICAL_HEADERS,
+                ('a', lambda headers: headers[0:1] + headers[3:5]),
+            ),
+            (StepAction.VERIFY_GAPS, (((2, 3),), 6)),
+        ),
+        # Another of ^
+        (
+            (StepAction.PERSIST_HEADERS, ('b', lambda headers: headers[:1])),
+            (StepAction.PERSIST_CHECKPOINT, 2),
+            (StepAction.VERIFY_GAPS, (((1, 1),), 3)),
+            (StepAction.VERIFY_PERSIST_RAISES, (
+                'b',
+                CheckpointsMustBeCanonical,
+                lambda headers: headers[1:3],
+            )),
+            (StepAction.PERSIST_HEADERS, ('a', lambda headers: headers[:1])),
+            (StepAction.VERIFY_GAPS, ((), 3)),
+            (StepAction.VERIFY_CANONICAL_HEADERS, ('a', lambda headers: headers[:2])),
+        ),
+        # Checkpoint a child of an uncle
+        (
+            (StepAction.PERSIST_HEADERS, ('b', lambda headers: headers[:1])),
+            (StepAction.VERIFY_GAPS, ((), 2)),
+            (StepAction.PERSIST_CHECKPOINT, 2),
+            # Persisting a checkpoint where the parent is not a match should cause the
+            # parent to become non-canonical.
+            (StepAction.VERIFY_GAPS, (((1, 1), ), 3)),
+            (StepAction.VERIFY_CANONICAL_HEAD, 2),
+            # Fill the uncled gap, the checkpoint (no-op) and the tail
+            (StepAction.PERSIST_HEADERS, ('a', lambda headers: headers)),
+            # Verify that block 1 gets canonicalized as A
+            (StepAction.VERIFY_CANONICAL_HEADERS, ('a', lambda headers: headers)),
+            (StepAction.VERIFY_GAPS, ((), 11)),
+        ),
+        # Don't override a checkpoint by persisting a chain that's a child of a gap
+        (
+            (StepAction.PERSIST_HEADERS, ('b', lambda headers: headers[:1])),
+            (StepAction.PERSIST_CHECKPOINT, 2),
+            (StepAction.VERIFY_GAPS, (((1, 1), ), 3)),
+            (StepAction.VERIFY_PERSIST_RAISES, (
+                'b',
+                CheckpointsMustBeCanonical,
+                lambda headers: headers[1:3]
+            )),
+        ),
+        # checkpointing should canonicalize matching parent header chain
+        (
+            (StepAction.PERSIST_HEADERS, ('a', lambda headers: headers[:2])),
+            (StepAction.PERSIST_HEADERS, ('b', lambda headers: headers[:3])),
+            (StepAction.PERSIST_CHECKPOINT, 3),
+            (StepAction.VERIFY_GAPS, ((), 4)),
+            (StepAction.VERIFY_CANONICAL_HEADERS, ('a', lambda headers: headers[:3])),
+        ),
```

Comment:
Wow! Quite a bunch of new test cases. Guess your intuition was right. Sorry, for leaving this half baked, I did not see all these possible special cases coming..